java-frontend/src/main/java/org/sonar/java/ExecutionTimeReport.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java;
+
+import java.time.Clock;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.stream.Collectors;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+public class ExecutionTimeReport {
+  private static final Logger LOG = Loggers.get(ExecutionTimeReport.class);
+
+  private static final int MAX_REPORTED_FILES = 10;
+  private static final long MIN_REPORTED_ANALYSIS_TIME_MS = 1000;
+  private static final long MIN_TOTAL_ANALYSIS_TIME_TO_REPORT_MS = 20_000;
+
+  private static class ExecutionTime {
+    private final String file;
+    private final long analysisTime;
+
+    public ExecutionTime(String file, long analysisTime) {
+      this.file = file;
+      this.analysisTime = analysisTime;
+    }
+  }
+
+  private static final Comparator<ExecutionTime> ORDER_BY_ANALYSIS_TIME_DESCENDING_AND_FILE_ASCENDING = (a, b) -> {
+    int compare = -Long.compare(a.analysisTime, b.analysisTime);
```

Comment:
why making it minus? if you know that b is after, just invert args.

Reply:
I started by inverting the args and then thought it would be more explicit to keep the same order in all this method and materialize the `descending` order by a minus. But, apparently, it generates more questions than answers, so ok, I'm inverting the args.